langkit/compiled_types.py

```diff
@@ -3626,6 +3653,113 @@ def require_hash_function(self):
+        # DSL API.
+        super(IteratorType, self).__init__(
+            name=name, is_ptr=True,
+            is_refcounted=True,
+            element_type=element_type,
+            has_equivalent_function=False,
+            hashable=False,
+            exposed=True)
+        CompiledTypeRepo.iterator_types.append(self)
+
+    @property
+    def name(self):
+        return self.element_type.name + names.Name('Iterator_Access')
+
+    @property
+    def api_name(self):
+        """
+        Name of the public iterator type.
+        """
+        return self.element_type.api_name + names.Name('Iterator')
+
+    @property
+    def py_converter(self):
+        """
+        Name of the Python class used to convert back and forth between
+        user-facing values (iterators) and C API values (pointers to iterator
+        records).
+
+        :rtype: str
+        """
+        return '_{}Converter'.format(self.api_name.camel)
+
+    def c_type(self, c_api_settings):
+        if (self.element_type.is_entity_type and
+                not self.element_type.emit_c_type):
+            return T.entity.iterator.c_type(c_api_settings)
+        else:
+            return CAPIType(c_api_settings, self.api_name)
+
+    def c_next(self, capi):
+        """
+        Name of the C API function to get the next value out of the iterator.
+
+        :param langkit.c_api.CAPISettings capi: Settings for the C API.
+        :rtype: str
+        """
+        return capi.get_name(self.api_name + names.Name('Next'))
+
+    def c_inc_ref(self, capi):
+        """
+        Name of the C API function to inc-ref an iterator value.
+
+        :param langkit.c_api.CAPISettings capi: Settings for the C API.
+        :rtype: str
+        """
+        return capi.get_name(self.api_name + names.Name('Inc_Ref'))
+
+    def c_dec_ref(self, capi):
+        """
+        Name of the C API function to dec-ref an iterator value.
+
+        :param langkit.c_api.CAPISettings capi: Settings for the C API.
+        :rtype: str
+        """
+        return capi.get_name(self.api_name + names.Name('Dec_Ref'))
+
+    @property
+    def to_public_converter(self):
+        return names.Name('To_Public') + self.api_name
+
+    @property
+    def emit_c_type(self):
+        """
+        Return whether to emit a C type for this type.
+
+        See StructType.emit_c_type.
+
+        :rtype: bool
+        """
+        return (not self.element_type.is_struct_type or
+                self.element_type.emit_c_type)
+
+    @property
+    def element_array_type_name(self):
+        return self.element_type.array.name
+
+    @property
+    def iterator_type_name(self):
+        """
+        Name of the Ada iterator type.
+
+        :rtype: names.Name
+        """
+        return (self.api_name
+                if self.is_string_type else
+                (names.Name('Internal') +
+                    self.element_type.name +
+                    names.Name('Iterator')))
```

Comment:
While this special treatment makes sense in `ArrayType.array_type_name` (we want to use `Wide_Wide_String` instead of declaring our own `Wide_Wide_Character` array), I see no reason for iterator not to behave similarly. Can you add a comment to explain this if it’s indeed required, or always return the `Internal_*_Iterator` type name?